fuzz/fuzz_targets/differential-new.rs

```diff
@@ -0,0 +1,85 @@
+#![no_main]
+
+use libfuzzer_sys::arbitrary::{Result, Unstructured};
+use libfuzzer_sys::fuzz_target;
+use wasmtime_fuzzing::generators::{DiffValue, SingleInstModule};
+use wasmtime_fuzzing::oracles::engine::{get_exported_function_signatures, DiffEngine};
+use wasmtime_fuzzing::oracles::{diff_wasmtime, differential, engine};
+
+const NUM_INVOCATIONS: usize = 5;
+
+fuzz_target!(|data: [u8; 1024]| {
+    // errors in `run` have to do with not enough input in `data`, which we
+    // ignore here since it doesn't affect how we'd like to fuzz.
+    drop(run(&data));
+});
+
+fn run(data: &[u8]) -> anyhow::Result<()> {
+    let mut u = Unstructured::new(data);
+
+    // Generate the Wasm module. TODO eventually, this should pick between the
+    // single-instruction and wasm-smith modules, but currently the wasm-smith
+    // module generation will eat up all of the random data, leaving none for
+    // the remaining decisions that follow (e.g., choosing an engine, generating
+    // arguments).
+    let module: &SingleInstModule = u.arbitrary()?;
+    let wasm = module.to_bytes();
+    let features = module.to_features();
+
+    // Choose a right-hand side Wasm engine--this will always be Wasmtime. The
+    // order (execute `lhs` first, then `rhs`) is important because, in some
+    // cases (e.g., OCaml spec interpreter), both sides register signal
+    // handlers; Wasmtime uses these signal handlers for catching various
+    // WebAssembly failures. On certain OSes (e.g. Linux x86_64), the signal
+    // handlers interfere, observable as an uncaught `SIGSEGV`--not even caught
+    // by libFuzzer. By always running Wasmtime second, its signal handlers are
+    // registered most recently and they catch failures appropriately. We create
+    // `rhs` first, however, so we have the option of creating a compatible
+    // Wasmtime engine (e.g., pooling allocator memory differences).
+    let rhs = diff_wasmtime::WasmtimeEngine::arbitrary_with_features(&mut u, &features)
+        .expect("unable to build a Wasmtime `rhs` engine");
+
+    // Choose a left-hand side Wasm engine.
+    let lhs = engine::choose(&mut u, &features, &rhs)
+        .expect("unable to choose an engine to fuzz against");
+
+    // Instantiate each engine and try each exported functions with various
+    // values.
+    let mut lhs_instance = lhs
+        .instantiate(&module.to_bytes())
+        .expect("failed to instantiate `lhs` module");
+    let mut rhs_instance = rhs
+        .instantiate(&module.to_bytes())
+        .expect("failed to instantiate `rhs` module");
+    for (name, signature) in get_exported_function_signatures(&wasm)
```

Comment:
One thing that may be helpful here is to use the concrete Wasmtime rhs to use Wasmtime's embedding API to determine this rather than using the `wasmparser` crate. 

I also think it's reasonable for the `differential` function itself to just force the rhs to the Wasmtime concrete type because I don't think there's much value in having two 100% generic engines since we always want one to be Wasmtime anyway.

Reply:
I don't see how this would change things much: we still want a separate function to gather up the signatures of the functions to execute. `wasmparser` is already a dependency and it doesn't seem to me that switching to Wasmtime would reduce much code (?).

Reply:
Using the Wasmtime embedding API is much more natural in my opinion, and the Wasmtime API is much more stable than the `wasmparser` API. I don't really want to keep up with random `wasmparser` changes way over here in the fuzzing code when one of the engines we're fuzzing against should always be Wasmtime anyway.

(I also don't think it's useful to be so abstract here that we could fuzz v8 against wasmi here, I would rather have "here's the wasmtime config" and "here's the other engine config" and the fuzzing is entirely driven by the Wasmtime embedding API plus requests to the engine config through a trait)